File: SorokinStepan/lw6/lw6/Sorter.cpp
#include "stdafx.h"
#include "Sorter.h"

Sorter::Sorter(AsyncManager & manager)
	: m_manager(manager)
	, m_threadCount(1)
{
}

Sorter::~Sorter()
{
}

void Sorter::SetArray(std::vector<int> const & array)
{
	m_elements.clear();
	for (auto &it : array)
	{
		m_elements.push_back(std::vector<int>());
		m_elements.back().push_back(it);
	}
}

void Sorter::StartSort()
{
	m_manager.SetThreadCount(m_threadCount);
	m_manager.Run();
	size_t blockSize = size_t(ceil(double(m_elements.size()) / m_threadCount));
	size_t targetPos = 0;
	std::mutex resultVecMutex;
	std::vector<std::vector<int>> resultVec;
	while (targetPos < m_elements.size())
	{
		std::vector<std::vector<int>> tempVec;
		if (targetPos + blockSize < m_elements.size())
		{
			tempVec = std::vector<std::vector<int>>(m_elements.begin() + targetPos, m_elements.begin() + targetPos + blockSize);
		}
		else
		{
			tempVec = std::vector<std::vector<int>>(m_elements.begin() + targetPos, m_elements.end());
		}
		targetPos += blockSize;
		m_manager.Post(boost::bind(Sorter::SortVec, tempVec, std::ref(resultVec), &resultVecMutex));
	}

	while(resultVec.size() < size_t(m_threadCount))
	{
		std::this_thread::sleep_for(std::chrono::milliseconds(10));
	}
	m_elements = std::move(resultVec);
	m_manager.Post(boost::bind(Sorter::SortVec, m_elements, std::ref(resultVec), &resultVecMutex));
	while (resultVec.empty())
	{
		std::this_thread::sleep_for(std::chrono::milliseconds(10));
	}
	m_elements = std::move(resultVec);
	m_manager.Stop();
}

void Sorter::SetThreadCount(int count)
{
	if (m_threadCount < 0)
	{
		m_threadCount = 1;
	}
	m_threadCount = count;
	m_manager.SetThreadCount(m_threadCount);
}

void Sorter::SortVec(std::vector<std::vector<int>>& vecToSort, std::vector<std::vector<int>>& resultVec, std::mutex* vecMutex)
{
	std::mutex tempMutex;
	while (vecToSort.size() > 1)
	{
		std::vector<std::vector<int>>  tempVec;
		size_t targetSize = size_t(ceil(double(vecToSort.size()) / 2.0));
		while (tempVec.size() < targetSize)
		{
			if (!vecToSort.empty())
			{
				if (vecToSort.size() > 1)
				{
					auto vec1 = vecToSort.back();
					vecToSort.pop_back();
					auto vec2 = vecToSort.back();
					vecToSort.pop_back();
					MergeVectors(vec1, vec2, std::ref(tempVec), &tempMutex);
				}
				else
				{
					auto vec1 = vecToSort.back();
					vecToSort.pop_back();
					tempMutex.lock();
					tempVec.push_back(vec1);
					tempMutex.unlock();
				}
			}
		}
		vecToSort = tempVec;
	}
	while (!tempMutex.try_lock())
	{
		std::this_thread::sleep_for(std::chrono::milliseconds(10));
	}
	tempMutex.unlock();

	vecMutex->lock();
	resultVec.push_back(vecToSort.front());
	vecMutex->unlock();
}

void Sorter::MergeVectors(std::vector<int> const& vec1, std::vector<int> const& vec2, std::vector<std::vector<int>> & mainVec, std::mutex * mutex)
{
	std::vector<int> resultVec;
	auto v1 = vec1;
	auto v2 = vec2;
	std::reverse(v1.begin(), v1.end());
	std::reverse(v2.begin(), v2.end());
	for (; !v1.empty() && !v2.empty();)
	{
		if (v1.back() > v2.back())
		{
			resultVec.push_back(v2.back());
			v2.pop_back();
		}
		else
		{
			resultVec.push_back(v1.back());
			v1.pop_back();
		}
	}
	while (!v1.empty())
	{
		resultVec.push_back(v1.back());
		v1.pop_back();
	}
	while (!v2.empty())
	{
		resultVec.push_back(v2.back());
		v2.pop_back();
	}
	mutex->lock();
	mainVec.push_back(std::move(resultVec));
	mutex->unlock();
}

File: SorokinStepan/lab7/lab7/Sorter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows;
using System.Threading;
using System.Threading.Tasks;

using Matrix = System.Collections.Generic.List<System.Collections.Generic.List<int>>;

namespace lab7
{
    class Sorter
    {

        public void SetArray(Matrix array)
        {
            m_array = array;
        }

        public void SetThreadCount(int count)
        {
            m_threadCount = count;
        }

        public List<int> StartSort()
        {
            int blockSize = (int)Math.Ceiling((double)m_array.Count() / (double)m_threadCount);
            int targetPos = 0;
            Matrix resultVec = new Matrix();

            Task[] tasks = new Task[m_threadCount];
            int taskCount = 0;

            while (targetPos < m_array.Count())
            {
                Matrix tempVec;
                if (targetPos + blockSize < m_array.Count())
                {
                    tempVec = m_array.GetRange(targetPos, blockSize);
                }
                else
                {
                    tempVec = m_array.GetRange(targetPos, m_array.Count() - targetPos);
                }
                targetPos += blockSize;
                tasks[taskCount++] = Task.Factory.StartNew(() => SortVec(tempVec, resultVec));
            }
            Task.WaitAll(tasks);
            SortVec(resultVec, m_array);
            return m_array[m_array.Count() - 1];
        }

        public void SortVec(Matrix vecToSort, Matrix resultVec)
        {
            while (vecToSort.Count() > 1)
            {
                Matrix tempVec = new Matrix();
                int targetSize = (int)(Math.Ceiling((double)(vecToSort.Count()) / 2.0));
                while (tempVec.Count() < targetSize)
                {
                    if (vecToSort.Count() != 0)
                    {
                        if (vecToSort.Count() > 1)
                        {
                            int size = vecToSort.Count();
                            List<int> vec1 = vecToSort[size - 1];
                            vecToSort.RemoveRange(size - 1, 1);
                            size = vecToSort.Count();
                            List<int> vec2 = vecToSort[size - 1];
                            vecToSort.RemoveRange(size - 1, 1);
                            MergeVectors(vec1, vec2, tempVec);
                        }
                        else
                        {
                            tempVec.Add(vecToSort[vecToSort.Count() - 1]);
                        }
                    }
                }
                vecToSort = tempVec;
            }

            resultVec.Add(vecToSort[0]);
        }

        public void MergeVectors(List<int> vec1, List<int> vec2, Matrix mainVec)
        {

            List<int> resultVec = new List<int>();
            List<int> v1 = vec1;
            List<int> v2 = vec2;
            v1.Reverse();
            v2.Reverse();
            
            while(v1.Count() != 0 && v2.Count() != 0)
            {
                if (v1[v1.Count() - 1] > v2[v2.Count() - 1])
                {
                    resultVec.Add(v2[v2.Count() - 1]);
                    v2.RemoveAt(v2.Count() - 1);
                }
                else
                {
                    resultVec.Add(v1[v1.Count() - 1]);
                    v1.RemoveAt(v1.Count() - 1);
                }
            }
            resultVec.AddRange(v1);
            resultVec.AddRange(v2);
            mainVec.Add(resultVec);
        }

        private Matrix m_array;
        private int m_threadCount;
    }
}
